File: sdk/js/src/protocol.js
/**
 * QuicFrame binary protocol codec — browser-side.
 *
 * Wire format (identical to the Go server):
 *
 *   ┌──────────────────┐
 *   │  FrameLen  (4B)  │  ← big-endian Uint32: byte count of Type + Payload
 *   ├──────────────────┤
 *   │  FrameType (1B)  │
 *   ├──────────────────┤
 *   │  Payload (msgpack│  ← @msgpack/msgpack encoded object
 *   └──────────────────┘
 *
 * Serialisation: MsgPack exclusively — never JSON.
 */

import { encode, decode } from '@msgpack/msgpack';

// ── Frame type discriminants ─────────────────────────────────────────────────
export const FRAME_REQUEST     = 0x01;
export const FRAME_RESPONSE    = 0x02;
export const FRAME_STREAM_DATA = 0x03;
export const FRAME_STREAM_END  = 0x04;
export const FRAME_ERROR       = 0x05;
export const FRAME_PING        = 0x06;
export const FRAME_PONG        = 0x07;

/** Maximum single frame size (64 MiB). */
export const MAX_FRAME_SIZE = 64 * 1024 * 1024;

// ── Encoder ──────────────────────────────────────────────────────────────────

/**
 * Encode a payload as a length-prefixed QuicFrame wire frame.
 *
 * @param {number} frameType
 * @param {object} payload
 * @returns {Uint8Array}
 */
export function encodeFrame(frameType, payload) {
  const encoded = encode(payload);                         // msgpack
  const frameLen = 1 + encoded.byteLength;

  if (frameLen > MAX_FRAME_SIZE) {
    throw new Error(`quicframe: payload too large (${frameLen} bytes)`);
  }

  const buf  = new ArrayBuffer(4 + frameLen);
  const view = new DataView(buf);
  view.setUint32(0, frameLen, false);   // big-endian length
  view.setUint8(4, frameType);          // type byte
  new Uint8Array(buf, 5).set(encoded);  // msgpack payload

  return new Uint8Array(buf);
}

// ── Decoder ──────────────────────────────────────────────────────────────────

/**
 * Stateful byte-buffer that accumulates raw bytes from a ReadableStream
 * and allows extracting complete frames.
 */
export class FrameBuffer {
  constructor() {
    /** @type {Uint8Array} */
    this._buf = new Uint8Array(0);
  }

  /** Append a new chunk from the stream reader. */
  push(chunk) {
    const merged = new Uint8Array(this._buf.byteLength + chunk.byteLength);
    merged.set(this._buf);
    merged.set(chunk, this._buf.byteLength);
    this._buf = merged;
  }

  /**
   * Try to extract one complete frame.
   * Returns `null` when not enough bytes have been buffered yet.
   *
   * @returns {{ frameType: number, payload: any } | null}
   */
  tryRead() {
    if (this._buf.byteLength < 4) return null;

    const view      = new DataView(this._buf.buffer, this._buf.byteOffset);
    const frameLen  = view.getUint32(0, false);           // big-endian
    const totalSize = 4 + frameLen;

    if (this._buf.byteLength < totalSize) return null;    // incomplete

    const frameType = this._buf[4];
    const msgpackPayload = this._buf.slice(5, totalSize);
    this._buf = this._buf.slice(totalSize);               // advance buffer

    return { frameType, payload: decode(msgpackPayload) };
  }
}

// ── Convenience constructors ─────────────────────────────────────────────────

/** @returns {Uint8Array} */
export function encodeRequest({ id, method, path, headers = {}, body = new Uint8Array(0), stream = false }) {
  return encodeFrame(FRAME_REQUEST, { id, method, path, headers, body, stream });
}

/** @returns {Uint8Array} */
export function encodePing() {
  return encodeFrame(FRAME_PING, { ts: Date.now() * 1_000_000 }); // nanoseconds
}

File: sdk/rust/src/protocol.rs
//! Wire-level frame codec — mirrors the Go server's `protocol` package.
//!
//! Frame wire format:
//! ```text
//! ┌──────────────────┐
//! │  FrameLen  (4B)  │  ← big-endian u32: byte count of Type + Payload
//! ├──────────────────┤
//! │  FrameType (1B)  │
//! ├──────────────────┤
//! │  Payload (msgpack│  ← rmp_serde-encoded struct
//! └──────────────────┘
//! ```

use std::collections::HashMap;

use bytes::{BufMut, BytesMut};
use serde::{Deserialize, Serialize};
use tokio::io::{AsyncRead, AsyncReadExt, AsyncWrite, AsyncWriteExt};

use crate::error::QfError;

// ── Frame type discriminants ─────────────────────────────────────────────────

pub const FRAME_REQUEST: u8 = 0x01;
pub const FRAME_RESPONSE: u8 = 0x02;
pub const FRAME_STREAM_DATA: u8 = 0x03;
pub const FRAME_STREAM_END: u8 = 0x04;
pub const FRAME_ERROR: u8 = 0x05;
pub const FRAME_PING: u8 = 0x06;
pub const FRAME_PONG: u8 = 0x07;

/// Maximum single frame size (64 MiB).
pub const MAX_FRAME_SIZE: u32 = 64 * 1024 * 1024;

// ── Payload types ────────────────────────────────────────────────────────────

/// Outgoing request (client → server).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Request {
    pub id: String,
    pub method: String,
    pub path: String,
    #[serde(default, skip_serializing_if = "HashMap::is_empty")]
    pub headers: HashMap<String, String>,
    #[serde(default, with = "serde_bytes", skip_serializing_if = "Vec::is_empty")]
    pub body: Vec<u8>,
    #[serde(default, skip_serializing_if = "std::ops::Not::not")]
    pub stream: bool,
}

/// Incoming response (server → client).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Response {
    pub id: String,
    pub status: i32,
    #[serde(default, skip_serializing_if = "HashMap::is_empty")]
    pub headers: HashMap<String, String>,
    #[serde(default, with = "serde_bytes", skip_serializing_if = "Vec::is_empty")]
    pub body: Vec<u8>,
    /// true when FrameStreamData/FrameStreamEnd frames follow.
    #[serde(default)]
    pub stream: bool,
}

/// A chunk in a streaming response.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StreamChunk {
    pub id: String,
    pub seq: u64,
    #[serde(default, with = "serde_bytes", skip_serializing_if = "Vec::is_empty")]
    pub data: Vec<u8>,
    pub r#final: bool,
}

/// Protocol-level error.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ErrorFrame {
    #[serde(default)]
    pub id: String,
    pub code: i32,
    pub message: String,
}

/// Keep-alive ping/pong.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PingFrame {
    pub ts: i64,
}

// ── Codec ────────────────────────────────────────────────────────────────────

/// Write a length-prefixed frame to an async writer.
///
/// ```text
/// [4B BE u32 frameLen][1B frameType][msgpack(payload)]
/// ```
pub async fn write_frame<W, T>(w: &mut W, frame_type: u8, payload: &T) -> Result<(), QfError>
where
    W: AsyncWrite + Unpin,
    T: Serialize,
{
    let encoded = rmp_serde::to_vec_named(payload)
        .map_err(|e| QfError::Codec(format!("msgpack encode: {e}")))?;

    let frame_len = 1u32 + encoded.len() as u32;
    if frame_len > MAX_FRAME_SIZE {
        return Err(QfError::Codec(format!(
            "payload too large: {frame_len} bytes"
        )));
    }

    let mut buf = BytesMut::with_capacity(4 + 1 + encoded.len());
    buf.put_u32(frame_len); // big-endian length
    buf.put_u8(frame_type); // type discriminant
    buf.extend_from_slice(&encoded); // msgpack payload

    w.write_all(&buf).await.map_err(QfError::Io)?;
    Ok(())
}

/// Decoded frame returned by [`read_frame`].
pub struct RawFrame {
    pub frame_type: u8,
    pub payload: Vec<u8>,
}

/// Read one length-prefixed frame from an async reader.
pub async fn read_frame<R>(r: &mut R) -> Result<RawFrame, QfError>
where
    R: AsyncRead + Unpin,
{
    // Read 4-byte big-endian length prefix.
    let mut header = [0u8; 4];
    r.read_exact(&mut header).await.map_err(QfError::Io)?;
    let frame_len = u32::from_be_bytes(header);

    if frame_len == 0 || frame_len > MAX_FRAME_SIZE {
        return Err(QfError::Codec(format!("invalid frame length: {frame_len}")));
    }

    let mut buf = vec![0u8; frame_len as usize];
    r.read_exact(&mut buf).await.map_err(QfError::Io)?;

    Ok(RawFrame {
        frame_type: buf[0],
        payload: buf[1..].to_vec(),
    })
}

/// Decode a [`Response`] from a raw msgpack payload.
pub fn decode_response(data: &[u8]) -> Result<Response, QfError> {
    rmp_serde::from_slice(data).map_err(|e| QfError::Codec(format!("decode response: {e}")))
}

/// Decode a [`StreamChunk`] from a raw msgpack payload.
pub fn decode_stream_chunk(data: &[u8]) -> Result<StreamChunk, QfError> {
    rmp_serde::from_slice(data).map_err(|e| QfError::Codec(format!("decode stream chunk: {e}")))
}

/// Decode an [`ErrorFrame`] from a raw msgpack payload.
pub fn decode_error(data: &[u8]) -> Result<ErrorFrame, QfError> {
    rmp_serde::from_slice(data).map_err(|e| QfError::Codec(format!("decode error frame: {e}")))
}
